File: bindings/java/src/main/java/com/dyber/pqc/Algorithm.java
/*
 * libpqc-dyber - Post-Quantum Cryptography Library
 * Copyright (c) 2024-2026 Dyber, Inc.
 * SPDX-License-Identifier: Apache-2.0 OR MIT
 */

package com.dyber.pqc;

/**
 * Registry of well-known algorithm name constants.
 *
 * <p>These match the C-level {@code PQC_KEM_*} and {@code PQC_SIG_*} defines.
 * You may also pass any algorithm name string directly.</p>
 */
public final class Algorithm {

    private Algorithm() {
        // non-instantiable
    }

    // ---- KEM algorithms ----

    public static final String ML_KEM_512 = "ML-KEM-512";
    public static final String ML_KEM_768 = "ML-KEM-768";
    public static final String ML_KEM_1024 = "ML-KEM-1024";

    public static final String HQC_128 = "HQC-128";
    public static final String HQC_192 = "HQC-192";
    public static final String HQC_256 = "HQC-256";

    public static final String BIKE_L1 = "BIKE-L1";
    public static final String BIKE_L3 = "BIKE-L3";
    public static final String BIKE_L5 = "BIKE-L5";

    public static final String FRODO_640_AES = "FrodoKEM-640-AES";
    public static final String FRODO_640_SHAKE = "FrodoKEM-640-SHAKE";
    public static final String FRODO_976_AES = "FrodoKEM-976-AES";
    public static final String FRODO_976_SHAKE = "FrodoKEM-976-SHAKE";
    public static final String FRODO_1344_AES = "FrodoKEM-1344-AES";
    public static final String FRODO_1344_SHAKE = "FrodoKEM-1344-SHAKE";

    // Hybrid KEM
    public static final String ML_KEM_768_X25519 = "ML-KEM-768+X25519";
    public static final String ML_KEM_1024_P256 = "ML-KEM-1024+P256";

    // ---- Signature algorithms ----

    public static final String ML_DSA_44 = "ML-DSA-44";
    public static final String ML_DSA_65 = "ML-DSA-65";
    public static final String ML_DSA_87 = "ML-DSA-87";

    public static final String SLH_DSA_SHA2_128S = "SLH-DSA-SHA2-128s";
    public static final String SLH_DSA_SHA2_128F = "SLH-DSA-SHA2-128f";
    public static final String SLH_DSA_SHA2_192S = "SLH-DSA-SHA2-192s";
    public static final String SLH_DSA_SHA2_192F = "SLH-DSA-SHA2-192f";
    public static final String SLH_DSA_SHA2_256S = "SLH-DSA-SHA2-256s";
    public static final String SLH_DSA_SHA2_256F = "SLH-DSA-SHA2-256f";

    public static final String FN_DSA_512 = "FN-DSA-512";
    public static final String FN_DSA_1024 = "FN-DSA-1024";

    public static final String MAYO_1 = "MAYO-1";
    public static final String MAYO_2 = "MAYO-2";
    public static final String MAYO_3 = "MAYO-3";
    public static final String MAYO_5 = "MAYO-5";

    // Stateful signatures
    public static final String LMS_SHA256_H10 = "LMS-SHA256-H10";
    public static final String LMS_SHA256_H15 = "LMS-SHA256-H15";
    public static final String LMS_SHA256_H20 = "LMS-SHA256-H20";
    public static final String LMS_SHA256_H25 = "LMS-SHA256-H25";

    public static final String XMSS_SHA2_10_256 = "XMSS-SHA2-10-256";
    public static final String XMSS_SHA2_16_256 = "XMSS-SHA2-16-256";
    public static final String XMSS_SHA2_20_256 = "XMSS-SHA2-20-256";

    // Hybrid signatures
    public static final String ML_DSA_65_ED25519 = "ML-DSA-65+Ed25519";
    public static final String ML_DSA_87_P256 = "ML-DSA-87+P256";
}

File: bindings/csharp/src/Algorithm.cs
// libpqc-dyber - Post-Quantum Cryptography Library
// Copyright (c) 2024-2026 Dyber, Inc.
// SPDX-License-Identifier: Apache-2.0 OR MIT

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Dyber.PQC
{
    /// <summary>
    /// Algorithm listing and query utilities.
    /// </summary>
    public static class Algorithm
    {
        // KEM algorithm constants
        public const string MlKem512 = "ML-KEM-512";
        public const string MlKem768 = "ML-KEM-768";
        public const string MlKem1024 = "ML-KEM-1024";
        public const string Hqc128 = "HQC-128";
        public const string Hqc192 = "HQC-192";
        public const string Hqc256 = "HQC-256";
        public const string BikeL1 = "BIKE-L1";
        public const string BikeL3 = "BIKE-L3";
        public const string BikeL5 = "BIKE-L5";

        // Signature algorithm constants
        public const string MlDsa44 = "ML-DSA-44";
        public const string MlDsa65 = "ML-DSA-65";
        public const string MlDsa87 = "ML-DSA-87";
        public const string SlhDsaSha2128s = "SLH-DSA-SHA2-128s";
        public const string SlhDsaSha2128f = "SLH-DSA-SHA2-128f";
        public const string FnDsa512 = "FN-DSA-512";
        public const string FnDsa1024 = "FN-DSA-1024";

        // Hybrid algorithm constants
        public const string MlKem768X25519 = "ML-KEM-768+X25519";
        public const string MlKem1024P256 = "ML-KEM-1024+P256";
        public const string MlDsa65Ed25519 = "ML-DSA-65+Ed25519";
        public const string MlDsa87P256 = "ML-DSA-87+P256";

        /// <summary>
        /// Returns a list of all enabled KEM algorithm names.
        /// </summary>
        public static IReadOnlyList<string> GetKemAlgorithms()
        {
            int count = NativeMethods.pqc_kem_algorithm_count();
            var list = new List<string>(count);
            for (int i = 0; i < count; i++)
            {
                IntPtr ptr = NativeMethods.pqc_kem_algorithm_name(i);
                if (ptr != IntPtr.Zero)
                    list.Add(Marshal.PtrToStringAnsi(ptr)!);
            }
            return list;
        }

        /// <summary>
        /// Returns a list of all enabled signature algorithm names.
        /// </summary>
        public static IReadOnlyList<string> GetSignatureAlgorithms()
        {
            int count = NativeMethods.pqc_sig_algorithm_count();
            var list = new List<string>(count);
            for (int i = 0; i < count; i++)
            {
                IntPtr ptr = NativeMethods.pqc_sig_algorithm_name(i);
                if (ptr != IntPtr.Zero)
                    list.Add(Marshal.PtrToStringAnsi(ptr)!);
            }
            return list;
        }

        /// <summary>
        /// Returns true if the given KEM algorithm is enabled.
        /// </summary>
        public static bool IsKemEnabled(string algorithm) =>
            NativeMethods.pqc_kem_is_enabled(algorithm) != 0;

        /// <summary>
        /// Returns true if the given signature algorithm is enabled.
        /// </summary>
        public static bool IsSignatureEnabled(string algorithm) =>
            NativeMethods.pqc_sig_is_enabled(algorithm) != 0;

        /// <summary>
        /// Returns the library version string.
        /// </summary>
        public static string GetVersion()
        {
            IntPtr ptr = NativeMethods.pqc_version();
            return Marshal.PtrToStringAnsi(ptr) ?? "unknown";
        }
    }
}
